sycl/include/CL/sycl/detail/backend_traits_level_zero.hpp

```diff
@@ -133,8 +137,25 @@ template <> struct BackendReturn<backend::ext_oneapi_level_zero, event> {
 template <> struct BackendInput<backend::ext_oneapi_level_zero, queue> {
   struct type {
     interop<backend::ext_oneapi_level_zero, queue>::type NativeHandle;
-    ext::oneapi::level_zero::ownership Ownership{
-        ext::oneapi::level_zero::ownership::transfer};
+    ext::oneapi::level_zero::ownership Ownership;
+
+    // TODO: Change this to be device when the deprecated constructor is
+    // removed.
+    std::shared_ptr<device_impl> Device;
```

Comment:
I see what you mean. Alternatively we could have this be a wrapper class around the pointer, allowing implicit conversion to and from `sycl::device`. Can you think of any problems that could cause?

Reply:
Would the following be possible?

* Add a private constructor to `device` that constructs a device as "not a device".
* Make the `BackendInput` struct a friend of the `device` class, so it can use that private constructor.
* Add a default member initializer for `Device` that calls this new `device` constructor.

Reply:
I am fairly confident that is possible, but I worry about such a device being accessible by the user. The user could create the type without a device and read the device from it, then pass that around. It may be an edge-case but I don't think it is worth figuring out all the places we would need to make special checks to make sure we're not given a "fake" device.

I was thinking of something along the lines of:
```c++
struct OptionalDevice {

  OptionalDevice() : DeviceImpl(nullptr) {}
  OptionalDevice(device dev) : DeviceImpl(getSyclObjImpl(dev)) {}

  operator device() const {
    if (!hasDevice())
      throw runtime_error("No device has been set.", PI_INVALID_DEVICE);
    return createSyclObjFromImpl<device>(DeviceImpl);
  }

  bool hasDevice() const { return DeviceImpl != nullptr; }
private:
  std::shared_ptr<device_impl> DeviceImpl;
};
```

That should allow most of the setting and getting cases, resulting in an exception if the device was not set before it was converted. I do however wonder how many drawbacks there potentially are to this approach.

Reply:
My concern with this approach is that it may be difficult to remove later.  Let's say we add `OptionalDevice` now.  Can we remove it later once the deprecation period is over for the old form of `make_queue`?  Once we remove the old form of `make_queue`, I'd like the struct to look like this:

```
struct {
  ze_command_queue_handle_t NativeHandle;
  device Device;
  ext::oneapi::level_zero::ownership Ownership{
      ext::oneapi::level_zero::ownership::transfer};
}
```

The approach I outline above with the private `device` constructor will let us get here eventually.  Will we be able to get here with the `OptionalDevice` approach?

Reply:
I have added the changes related to `OptionalDevice` to illustrate it. If we decide for the device constructor I will revert it.

I agree that having it as `device` is the end goal once we are rid of the deprecated API, but since the input type is unpacked in sycl/include/sycl/ext/oneapi/backend/level_zero.hpp where the relevant device is also finally found, the input type lives solely in the headers. As such we should not have any problems changing the member type when we remove the deprecated API.

Reply:
I think this approach is OK, but see my specific comments in the code.

I still think the API spec needs to document the deprecated API and the feature-test macro version should be incremented, as I noted in my other comment.